modules/core/04-channel/keeper/upgrade.go

```diff
@@ -109,6 +109,37 @@ func (k Keeper) WriteUpgradeTryChannel(
 	emitChannelUpgradeTryEvent(ctx, portID, channelID, channel, proposedUpgrade)
 }
 
+// WriteUpgradeOpenChannel writes the agreed upon upgrade fields to the channel fields. This can be called in one of two cases:
+// - In the UpgradeAck step of the handshake both sides have already flushed any in-flight packets.
+// - In the UpgradeOpen step of the handshake.
+func (k Keeper) WriteUpgradeOpenChannel(ctx sdk.Context, portID, channelID string) {
```

Comment:
Question. It would be nice if we could get information about who called us (`Ack` or `Open`) since that can be used in the error messages and when incrementing the telemetry counter. Would a bool (or enum :^)) be the preferred approach?

I don't think we can use the previousState here to make assumptions about who called us but I haven't thought too hard about it yet.

Reply:
why can't we use the `previousState`?

Reply:
If my understanding of the state transitions are correct (ref https://github.com/cosmos/ibc/pull/978/files). We can end up in the subprotocol from both open and from ack with channel state being `ACKUPGRADE` for channel `A`.

Reply:
are you talking about the case where this subprotocol is being called on chain B, where chain A has automatically moved from ACK -> OPEN bc the pending in flight packets check results in an automatic chan OPEN?

Reply:
wouldn't be be in `TRYUPGRADE`. To be a bit more verbose about my confusion, during `ChanUpgradeAck` and after `startFlushUpgradeHandshake` the states of `A` and `B` will be `(ACKUPGRADE, TRYUPGRADE)`. If at this point both A and B have finished flushing, `openUpgradeHandShake` will be called with those two states for A and B.

OTOH, if they have not finished flushing, after they do, the relayer will submit the `ChanUpgradeClose` messages to both `A` and `B` and the states, having not changed, will still be ``(ACKUPGRADE, TRYUPGRADE)`.

I feel I'm missing something here but can't see it yet. Either way, this doesn't affect the core logic, its mostly for the telemetry counter and the err messages.